velox/expression/tests/ExprOptimizerTest.cpp

```diff
@@ -0,0 +1,148 @@
+#include "velox/expression/ExprOptimizer.h"
+#include "velox/expression/ExprRewriteRegistry.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
+#include "velox/parse/Expressions.h"
+#include "velox/parse/ExpressionsParser.h"
+#include "velox/parse/TypeResolver.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+namespace facebook::velox::expression::test {
+
+class ExprOptimizerTest : public testing::Test,
+                          public velox::test::VectorTestBase {
+ protected:
+  static void SetUpTestCase() {
+    memory::MemoryManager::testingSetInstance(memory::MemoryManager::Options{});
+  }
+
+  void SetUp() override {
+    functions::prestosql::registerAllScalarFunctions("");
+    parse::registerTypeResolver();
+  }
+
+  void TearDown() override {
+    expression::ExprRewriteRegistry::instance().clear();
+  }
+
+  core::TypedExprPtr makeTypedExpr(
+      const std::string& expression,
+      const RowTypePtr& type) {
+    auto untyped = parse::parseExpr(expression, {});
+    return core::Expressions::inferTypes(untyped, type, execCtx_->pool());
+  }
+
+  core::TypedExprPtr optimize(
+      const std::string& expression,
+      const RowTypePtr& type = ROW({}),
+      bool replaceEvalErrorWithFailExpr = false) {
+    const auto expr = makeTypedExpr(expression, type);
+    return expression::optimize(
+        expr, queryCtx_.get(), pool(), replaceEvalErrorWithFailExpr);
+  }
+
+  void testExpression(
+      const std::string& input,
+      const std::string& expected,
+      const RowTypePtr& inputType = ROW({}),
+      const RowTypePtr& expectedType = ROW({})) {
+    auto optimized = optimize(input, inputType);
+    auto optimizedExpected = optimize(expected, expectedType);
+    // The constant value in ConstantTypedExpr can either be in valueVector_ or
+    // in the variant value_. String comparison is used to compare the optimized
+    // expressions, since ITypedExpr comparison with equality operator will fail
+    // for `ConstantTypedExpr`s if the underlying constant values do not have
+    // the same internal representation.
+    ASSERT_EQ(optimized->toString(), optimizedExpected->toString());
+  }
+
+  void testFailExpression(
+      const core::TypedExprPtr& expr,
+      const std::string& expectedError) {
+    ASSERT_TRUE(expr->isCallKind());
+    const auto call = expr->asUnchecked<core::CallTypedExpr>();
+    ASSERT_EQ(call->name(), kFail);
+
+    ASSERT_TRUE(call->inputs().at(0)->isConstantKind());
+    const auto error =
+        call->inputs().at(0)->asUnchecked<core::ConstantTypedExpr>();
+    ASSERT_TRUE(error->type()->isVarchar());
+    LOG(ERROR) << error->toString();
+    ASSERT_TRUE(error->toString().find(expectedError) != std::string::npos);
+  }
+
+  void testFailExpression(
+      const std::string& expression,
+      const std::string& expectedError) {
+    auto optimized = optimize(expression, ROW({}), true);
+    testFailExpression(optimized, expectedError);
```

Comment:
testFailExpression -> assertFailCall (since this method asserts that 'expr' is a fail function call)